File: lca_bt.py
# time Complexity: O(N)
# space Complexity: O(N)

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    pathP = []
    pathQ = []

    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        def helper(root, p, q, nums):
            # Base
            if root == None:
                return
            # Logic
            if root.val == p.val:
                temp = nums[:]
                self.pathP = temp
                self.pathP.append(p)
                self.pathP.append(p)
            if root.val == q.val:
                temp = nums[:]
                self.pathQ = temp
                self.pathQ.append(q)
                self.pathQ.append(q)

            nums.append(root)
            helper(root.left, p, q, nums)
            helper(root.right, p, q, nums)
            nums.pop()

        helper(root, p, q, [])
        for i in range(len(self.pathP)):
            if self.pathP[i] != self.pathQ[i]:
                return self.pathQ[i-1]

File: lca_bt.java
//TC: O(n)
//SC: O(h)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<TreeNode> pathp;
    List<TreeNode> pathq;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
       // this.pathp=new ArrayList<>();
      //  this.pathq=new ArrayList<>();
      helper(root,p,q,new ArrayList<>());
      for(int i=0;i<pathp.size();i++)
      {
        if(pathp.get(i)!=pathq.get(i))
        {
            return pathp.get(i-1);
        }
      }
return null;
        
    }
    private  void helper(TreeNode root, TreeNode p, TreeNode q,List<TreeNode> path)
    {
        if(root==null)return;
        path.add(root);
        if(root==p)
        {
             
            pathp=new ArrayList<>(path);
         pathp.add(root);

        }
         if(root==q)
        {
            pathq=new ArrayList<>(path);
            pathq.add(root);

        }


        helper(root.left,p,q,path);
        helper(root.right,p,q,path);
        path.remove(path.size()-1);
    }
}